sys/kern/pty.c

```diff
@@ -0,0 +1,277 @@
+#include <sys/filedesc.h>
+#include <sys/mutex.h>
+#include <sys/ringbuf.h>
+#include <sys/ttycom.h>
+#include <stdatomic.h>
+#include <stdio.h>
+#include <sys/condvar.h>
+#include <sys/mimiker.h>
+#include <sys/file.h>
+#include <sys/proc.h>
+#include <sys/fcntl.h>
+#include <sys/errno.h>
+#include <sys/stat.h>
+#include <sys/devfs.h>
+#include <sys/linker_set.h>
+#include <sys/tty.h>
+
+#define MAX_PTYS 16
+
+/* Must be greater than or equal to the length of the path to any slave device.
+ * Right now slave devices are at /dev/pts/<number>, so this should suffice. */
+#define PTY_PATH_MAX_LEN 15
+
+static devfs_node_t *pts_dir;
+
+typedef struct {
+  atomic_int pt_number; /* PTY number, if allocated. -1 means free. */
+  condvar_t pt_incv;    /* CV for readers */
+  condvar_t pt_outcv;   /* CV for writers */
+} pty_t;
+
+static pty_t pty_array[MAX_PTYS];
+
+static pty_t *pty_alloc(void) {
+  for (int i = 0; i < MAX_PTYS; i++) {
+    pty_t *pty = &pty_array[i];
+    if (atomic_compare_exchange_strong(&pty->pt_number, &(int){-1}, i + 1))
+      /* Got a free PTY. */
+      return pty;
+  }
+  return NULL;
+}
+
+static void pty_free(pty_t *pty) {
+  assert(pty->pt_number >= 0);
+  pty->pt_number = -1;
+}
+
+static int pty_read(file_t *f, uio_t *uio) {
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+  int error;
+  size_t start_resid = uio->uio_resid;
+
+  if (uio->uio_resid == 0)
+    return 0;
+
+  SCOPED_MTX_LOCK(&tty->t_lock);
+
+  /* Wait until there is at least one byte of data. */
+  while (ringbuf_empty(&tty->t_outq)) {
+    /* Don't wait for data if slave device isn't opened. */
+    if (!tty_opened(tty))
+      return 0;
+    if (cv_wait_intr(&pty->pt_incv, &tty->t_lock))
+      return ERESTARTSYS;
+  }
+
+  /* Data is available: transfer as much as we can. */
+  error = ringbuf_read(&tty->t_outq, uio);
+  tty_getc_done(tty);
+
+  /* Don't report errors on partial reads. */
+  if (start_resid > uio->uio_resid)
+    error = 0;
+
+  return error;
+}
+
+/* Write at single character to the master side of a pseudoterminal,
+ * i.e. to the input queue of the slave tty.
+ * If the input queue is full, sleep only if the slave tty has users. */
+static int pty_putc_sleep(tty_t *tty, pty_t *pty, uint8_t c) {
+  while (!tty_input(tty, c)) {
+    if (!tty_opened(tty))
+      return EIO;
+    if (cv_wait_intr(&pty->pt_outcv, &tty->t_lock))
+      return ERESTARTSYS;
+  }
+  return 0;
+}
+
+static int pty_write(file_t *f, uio_t *uio) {
+  tty_t *tty = f->f_data;
+  pty_t *pty = tty->t_data;
+  int error = 0;
+  uint8_t c;
+
+  if (uio->uio_resid == 0)
+    return 0;
+
+  size_t start_resid = uio->uio_resid;
+
+  SCOPED_MTX_LOCK(&tty->t_lock);
+
+  while (uio->uio_resid > 0) {
+    if ((error = uiomove(&c, 1, uio)))
+      break;
+    if ((error = pty_putc_sleep(tty, pty, c))) {
+      /* Undo the last uiomove(). */
+      uio->uio_resid++;
```

Comment:
IIRC `uio_t` fields are modified only in `uiomove` and I would like to keep it this way. Is there a simple way to reorganize this code not to break the premise ?

Reply:
To do so, we would need something like `pty_can_putc()` that checks if `tty_input()` on the slave device will succeed. The thing is, the outcome of `tty_input(c)` is heavily dependent on the value of `c` and terminal settings: e.g. when the `ISIG` flag is set, `tty_input('^C')` will always succeed, since the character isn't actually stored in the input queue. So IMO it's not possible. I also noticed that simply incrementing `uio_resid` is completely broken, as `uio_offset` and other fields need to be adjusted. I suggest adding a function for rolling back `uiomove()` operations, something like `void uio_rollback(uio_t *uio, size_t nbytes)`

Reply:
OK. I see... thanks for the explanation.